rfc/rfc-83/rfc-83.md

```diff
@@ -0,0 +1,212 @@
+
+We've already covered the updated logic flow of `HoodieIncrSource` and cloud events incremental source.
+By comparing them, we can come up with three abstract steps:
+- Get metadata
+- Fetch data
+- Post actions
+
+The following is a table to describe what `HoodieIncrSource` and cloud events incremental source need to do for each step.
+
+| *                        | Get Metadata                                                                                                                                                                               | Fetch Data                                                                         | Post Actions                                         |
+|--------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------|------------------------------------------------------|
+| `HoodieIncrSource`       | - Pass checkpoint and query config to `IncrementalQueryAnalyzer` and get `QueryContext`                                                                                                    | - Use `QueryContext` and `IncrementalRelation` to query data                       | - Filter the result<br/>- Report metrics<br/> - etc. |
+| Cloud Events Incr Source | - Pass checkpoint and query config to `IncrementalQueryAnalyzer` and get `QueryContext`<br/>- Fetch eligible `CloudObjectMetadata` from cloud metadata Hudi table<br/>- Filter the results | - Use `QueryContext` and `CloudObjectMetadata` to fetch objects from cloud storage | - Report metrics<br/>- etc.                          |
+
+Note that Flink's `IncrementalInputSplits` can also be abstracted into these flow, but it's not in the scope of this RFC.
+
+#### IncrementalQueryRunner
+With the abstraction above, we can come up with the interface `IncrementalQueryRunner`.
+```java
+public interface IncrementalQueryRunner<M, D> {
+  
+  M getMetadata();
+
+  D fetchData();
+
+  D postActions();
+}
+```
+We can have the following implementations based on this interface:
+- `HoodieSourceIncrementalQueryRunner`: Covers incremental logic in `HoodieIncrSource`
+- `CloudEventsSourceIncrementalQueryRunner`: Covers incremental logic in `S3EventsHoodieIncrSource` and `GCSEventsHoodieIncrSource`
+  - Noticing the `Get Metadata` phase of getting cloud events source incremental data follows the same flow of
+    `getMetadata` -> `fetchData` -> `postActions`. We can add a nested query runner `CloudMetadataIncrementalQueryRunner` 
+    to get cloud metadata
+- (not in scope)`InputSplitsIncrementalQueryRunner`: Covers incremental logic in `IncrementalInputSplits`
+
+The dependency graph is presented here:
+![dep_graph](./incrqrunner_dep_graph.png)
+
+
+## Rollout/Adoption Plan
+
+- What impact (if any) will there be on existing users?
```

Comment:
TBD

Reply:
Another critical change we made since 1.0 is the `InstantRange` start has been inclusive, while it is exclusive before, we may need to make it clear in this RFC.